File: python_src/12.整数转罗马数字.py
#
# @lc app=leetcode.cn id=12 lang=python3
#
# [12] 整数转罗马数字
#

# @lc code=start
class Solution:
    def intToRoman(self, num: int) -> str:
        mp = {0: '', 1: 'I', 5: 'V', 10: 'X', 50: 'L', 100: 'C', 500: 'D', 1000: 'M'}
        s = ''
        s_num = str(num)
        size = len(s_num)
        for i, digit in enumerate(s_num):
            n = int(digit)
            p = 10 ** (size - i - 1)
            base = n // 5
            tmp = offset = n % 5
            offset = mp[p] * offset if offset != 4 else mp[p] + mp[(n + 1) * p]
            base = mp[5 * base * p] * (0 if tmp == 4 else base)
            s += base + offset

        return s

# @lc code=end

s = Solution()
print(s.intToRoman(3749))
print(s.intToRoman(58))
print(s.intToRoman(1994))

File: rust_src/12.整数转罗马数字.rs
/*
 * @lc app=leetcode.cn id=12 lang=rust
 *
 * [12] 整数转罗马数字
 */

// @lc code=start
impl Solution {
    pub fn int_to_roman(num: i32) -> String {
		let mut i = num;
		let mut j = 0;
		let mut res = String::new();
		let arr = [["", "I", "II", "III", "IV", "V", "VI", "VII", "VIII", "IX"],
					["", "X", "XX", "XXX", "XL", "L", "LX", "LXX", "LXXX", "XC"],
					["" , "C", "CC", "CCC", "CD", "D", "DC", "DCC", "DCCC", "CM"],
					["", "M", "MM", "MMM", "", "", "", "", "", ""]];

		while i != 0
		{
			let digit = i % 10;
			i /= 10;
			res.insert_str(0, arr[j][digit as usize]);
			j += 1;
		}
		
		return res;
    }
}

// @lc code=end

